docs/user-guide/templates/pipeline-templates.md

```diff
@@ -0,0 +1,194 @@
+* `nodejs/test@1.0` - this will use the most recent version of `nodejs/test` than is less than 1.1.0. This is essentially the `latest` tag without crossing a minor version boundary.
+* `nodejs/test@1.0.4` - this is the most predictable way to specify a pipeline's behavior and is not affected by future changes to the template.
+
+#### Example
+
+For this configuration:
+```yaml
+template: nodejs/test@stable
+shared: 
+  environment:
+    FOO: bar
+```
+
+
+Screwdriver takes the template configuration and plugs it in, so that the `screwdriver.yaml` becomes:
+
+```yaml
+shared:
+  environment:
+    FOO: bar
+jobs:
+    main:
+      image: node:18
+      requires: [~pr, ~commit]
+      steps:
+        - install: npm install
+        - test: npm test
+      secrets:
+            - NPM_TOKEN
+```
+
+## Creating a template
+
+Publishing and running pipeline templates must be done from a Screwdriver pipeline.
+
+### Writing a template yaml
+
+To create a template, create a new repo with a `sd-template.yaml` file. The file should contain a namespace, name, version, description, maintainer email, and a config with jobs. Basic example can be found in our [screwdriver-cd-test/pipe;line-template-example repo](https://github.com/screwdriver-cd-test/pipeline-template-example).
+
+Example `sd-template.yaml`:
+
+```yaml
+namespace: myNamespace
+name: template_name
+version: '1.3'
+description: pipeline template for testing
+maintainer: foo@bar.com
+config:
+    jobs:
+      main:
+        requires: [~pr, ~commit]
+        steps:
+          - install: npm install
+          - test: npm test
+        environment:
+            FOO: bar
+        secrets:
+            - NPM_TOKEN
+```
+
+### Writing a screwdriver yaml for your template repo
+
+#### Validating templates
+
+To validate your template, run the `pipeline-template-validate` script from the `screwdriver-template-main` npm module in your `main` job to validate your template. This means the build image must have NodeJS and NPM properly installed to use it. To publish your template, run the `pipeline-template-publish` script from the same module in a separate job.
+
+By default, the file at `./sd-template.yaml` will be read. However, a user can specify a custom path using the env variable: `SD_TEMPLATE_PATH`.
+
+You can also validate your `sd-template.yaml` and `screwdriver.yaml` through the UI by copy pasting it at `<YOUR_UI_URL>/validator`.
+
+#### Tagging templates
+You can optionally tag a specific template version by running the `pipeline-template-tag` script from the `screwdriver-template-main` npm package. This must be done by the same pipeline that your template is created by. You will need to provide arguments to the script: template namespace, name and tag. You can optionally specify a version; the version needs to be an exact version (see `tag` step). If the version is omitted, the most recent version will be tagged (see `autotag` step).
```

Comment:
same here, please provide links for:
`pipeline-template-tag` script
`screwdriver-template-main` npm package

see `tag` step => see `tag` step in example below.

Reply:
One common theme I see here is the instructions for running script, I do see them in readme.md file https://github.com/screwdriver-cd/template-main?tab=readme-ov-file#validating-a-template-1 etc.

However, right thing here is to use `npx`. All instructions would become a one liner

` npx -y -p screwdriver-template-main pipeline-template-validate`

However this doesn't work today, because file permissions are messed up.

```
npx -y -p screwdriver-template-main pipeline-template-validate
sh: /Users/jithine/.npm/_npx/8e289187f01fda48/node_modules/.bin/pipeline-template-validate: Permission denied
❯ ls -al /Users/jithine/.npm/_npx/8e289187f01fda48/node_modules/screwdriver-template-main/bin
total 120
drwxr-xr-x  17 jithine  staff  544 Mar  1 12:23 .
drwxr-xr-x  11 jithine  staff  352 Mar  1 12:23 ..
-rwxr-xr-x   1 jithine  staff   88 Mar  1 12:23 getVersionFromTag.js
-rwxr-xr-x   1 jithine  staff  480 Mar  1 12:23 main.js
-rwxr-xr-x   1 jithine  staff  106 Mar  1 12:23 pipelineTemplateGetVersionFromTag.js
-rw-r--r--   1 jithine  staff   93 Mar  1 12:23 pipelineTemplatePublish.js
-rw-r--r--   1 jithine  staff   92 Mar  1 12:23 pipelineTemplateRemove.js
-rw-r--r--   1 jithine  staff   96 Mar  1 12:23 pipelineTemplateRemoveTag.js
-rw-r--r--   1 jithine  staff  100 Mar  1 12:23 pipelineTemplateRemoveVersion.js
-rw-r--r--   1 jithine  staff   89 Mar  1 12:23 pipelineTemplateTag.js
-rw-r--r--   1 jithine  staff   94 Mar  1 12:23 pipelineTemplateValidate.js
-rwxr-xr-x   1 jithine  staff   75 Mar  1 12:23 publish.js
❯ chmod u+x /Users/jithine/.npm/_npx/8e289187f01fda48/node_modules/screwdriver-template-main/bin/*
❯ npx -y -p screwdriver-template-main pipeline-template-validate
Error: ENOENT: no such file or directory, open './sd-template.yaml'
    at Object.openSync (node:fs:601:3)
    
```